crates/chia-consensus/src/multiprocess_validation.rs

```diff
@@ -0,0 +1,238 @@
+use chia_bls::aggregate_verify;
+use chia_bls::BlsCache;
+use chia_bls::PublicKey;
+use chia_protocol::Coin;
+use chia_protocol::Program;
+use chia_protocol::SpendBundle;
+use chia_traits::Streamable;
+use clvmr::{ENABLE_BLS_OPS_OUTSIDE_GUARD, ENABLE_FIXED_DIV};
+use std::sync::{Arc, Mutex};
+// use std::thread;
+use std::time::{Duration, Instant};
+
+// currently in multiprocess_validation.py
+// called via blockchain.py from full_node.py when a full node wants to add a block or batch of blocks
+pub fn pre_validate_blocks_multiprocessing() {
+    batch_pre_validate_blocks()
+}
+
+// currently in multiprocess_validation.py
+// called in threads from pre_validate_blocks_multiprocessing
+fn batch_pre_validate_blocks() {}
+
+// currently in mempool_manager.py
+// called in full_node.py when adding a transaction
+pub fn pre_validate_spendbundle(
+    new_spend: SpendBundle,
+    max_cost: u64,
+    constants: ConsensusConstants,
+    peak_height: u32,
+    syncing: bool,
+    cache: Arc<Mutex<BlsCache>>,
+) -> Result<OwnedSpendBundleConditions, ErrorCode> {
+    if new_spend.coin_spends.is_empty() {
+        Err(ErrorCode::InvalidSpendBundle)
+    } else {
+        let (result, _duration) = validate_clvm_and_signature(
+            &new_spend,
+            max_cost,
+            constants,
+            peak_height,
+            syncing,
+            cache,
+        )?;
+        Ok(result)
+    }
+}
+
+// currently in mempool_manager.py
+// called in threads from pre_validate_spend_bundle()
+// returns (error, cached_results, new_cache_entries, duration)
+fn validate_clvm_and_signature(
+    spend_bundle: &SpendBundle,
+    max_cost: u64,
+    constants: ConsensusConstants,
+    height: u32,
+    syncing: bool,
+    cache: Arc<Mutex<BlsCache>>,
+) -> Result<(OwnedSpendBundleConditions, Duration), ErrorCode> {
+    let start_time = Instant::now();
+    let additional_data = constants.agg_sig_me_additional_data;
+    let program: BlockGenerator = simple_solution_generator(spend_bundle)?;
+    let npcresult = get_name_puzzle_conditions(program, max_cost, true, height, &constants)?;
+    let (pks, msgs) = pkm_pairs(npcresult.clone(), additional_data.as_slice())?;
+    // Verify aggregated signature
+    if !{
+        if syncing {
+            // if we're syncing use the chia_bls::aggregate_verify to avoid using the cache
+            let data: Vec<(&PublicKey, &[u8])> = pks
+                .iter()
+                .zip(msgs.iter().map(|msg| msg.as_slice()))
+                .collect();
+            aggregate_verify(&spend_bundle.aggregated_signature, data)
+        } else {
+            // if we're fully synced then use the cache
+            cache
+                .lock()
+                .unwrap()
+                .aggregate_verify(pks, msgs, &spend_bundle.aggregated_signature)
+        }
+    } {
+        return Err(ErrorCode::InvalidSpendBundle);
+    }
+    Ok((npcresult, start_time.elapsed()))
+}
+
+// #[cfg(feature = "py-bindings")]
```

Comment:
Is this going to be uncommented/removed later?

Reply:
Done